IPs/RSKIP144.md

```diff
@@ -9,48 +9,88 @@
+## New block header field
 
-# Suggested miner implementation
+A new field `partitionEnds` is added to the block header. It determines how transactions are partitioned in a block.
+
+This field consists of an array of short unsigned integers that indicates at which position in the transaction list each partition ends. For example, in a block with 10 transactions, `partitionEnds = [3, 6]` indicates that the first _parallel partition_ contains transactions 0, 1 and 2; the second _parallel partition_ contains transactions 3, 4 and 5; and the _sequential partition_ contains transactions 6 to 9.
+
+The REMASC transaction must be included as the last transaction of the sequential partition.
+
+An empty `partitionEnds` indicates that all transactions go in the _sequential partition_.
+
+Values in `partitionsEnd` must be greater than 0 and in ascending order. An empty `partitionEnds` indicates that all transactions go in the sequential partition. The maximum number of parallel partitions that the miner can specify is equal to the minimum number of cores required to run the RSK node, this maximum is called  `maxTransactionExecutionThreads`.
+
+## New block validation consensus
+
+The new consensus mechanism allows to reject blocks that, when executed in parallel, can result in different outputs.
 
-When a miner executes transactions to create an (unsolved) block, the miner must execute the transactions serially and decide which partition the transaction belongs to. The block gas limit is replaced by a per-partition gas limit. We say that a transaction is *connected* to another transaction if:
+For simplicity, three types of connections between transactions are identified:
+- Both transactions are from the same sender account
+- Both transactions write the same storage key
+- One transaction reads a key that the other transaction writes
 
-1. Both transactions are from the same sender account
-2. Both transactions write the same storage key
-3. One transaction reads the same storage key that the other transaction writes
+Any pair of transactions with any of these characteristics are required to be in the same _parallel partition_ to avoid non-deterministic results, or one needs to be executed in the _sequential partition_
```

Comment:
Can be both in the sequential also. I would say that two connected transactions cannot be in different parallel partitions.